### ISSUE TITLE
Long prediction using KFServing API 504 Gateway Timeout

### ISSUE THREAD

**Issue by decewei** [kubeflow/kfserving#892]
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
Ran a custom container, using short/small data that will respond within 10-second is working fine.

However, we ran a custom container with a long input, that will take more than 30 seconds, the curl will get 504 Gateway Timeout. Queue-proxy has 127.0.0.1:8080 connection refused error.

**What did you expect to happen:**
Should wait for a response that takes longer.
Should be able to read probe.

**Environment:**
- Knative Version: 0.11.1
- KFServing Version: 0.2.2
- Kubeflow version: 1.0.0
- Microk8s: 1.18.3
- Kubernetes version: (use `kubectl version`): 1.18.3
- OS (e.g. from `/etc/os-release`): RHEL

**Comment by yuzisun** [comment]
@decewei currently the default timeout is 60s, can you paste the logs for the `queue-proxy` error?

**Comment by decewei** [comment]
Is there a way to change the timeout? 
Here:
```
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
aggressive probe error:  dial tcp 127.0.0.1:8080: connect: connection refused
{"level":"error","ts":"2020-06-19T18:43:56.428Z","logger":"queueproxy","caller":"network/error_handler.go:31","msg":"error reverse proxying request; sockstat: sockets: used 101897\nTCP: inuse 33 orphan 1 tw 642 alloc 2694 mem 115\nUDP: inuse 0 mem 8\nUDPLITE: inuse 0\nRAW: inuse 0\nFRAG: inuse 0 memory 0\n","commit":"804b3bb","knative.dev/key":"default/iae-timeline-predictor-default-pspwr","knative.dev/pod":"iae-timeline-predictor-default-pspwr-deployment-bc4f66df5-zhdlz","error":"context canceled","stacktrace":"knative.dev/serving/vendor/knative.dev/pkg/network.ErrorHandler.func1\n\tknative.dev/serving/vendor/knative.dev/pkg/network/error_handler.go:31\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\tnet/http/httputil/reverseproxy.go:251\nknative.dev/serving/pkg/queue.(*appRequestMetricsHandler).ServeHTTP\n\tknative.dev/serving/pkg/queue/request_metric.go:196\nmain.proxyHandler.func1\n\tknative.dev/serving/cmd/queue/main.go:169\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.ForwardedShimHandler.func1\n\tknative.dev/serving/pkg/queue/forwarded_shim.go:100\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.(*timeoutHandler).ServeHTTP.func1\n\tknative.dev/serving/pkg/queue/timeout.go:80"}
{"level":"error","ts":"2020-06-19T18:44:56.472Z","logger":"queueproxy","caller":"network/error_handler.go:31","msg":"error reverse proxying request; sockstat: sockets: used 101871\nTCP: inuse 62 orphan 1 tw 646 alloc 2721 mem 109\nUDP: inuse 0 mem 8\nUDPLITE: inuse 0\nRAW: inuse 0\nFRAG: inuse 0 memory 0\n","commit":"804b3bb","knative.dev/key":"default/iae-timeline-predictor-default-pspwr","knative.dev/pod":"iae-timeline-predictor-default-pspwr-deployment-bc4f66df5-zhdlz","error":"context canceled","stacktrace":"knative.dev/serving/vendor/knative.dev/pkg/network.ErrorHandler.func1\n\tknative.dev/serving/vendor/knative.dev/pkg/network/error_handler.go:31\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\tnet/http/httputil/reverseproxy.go:251\nknative.dev/serving/pkg/queue.(*appRequestMetricsHandler).ServeHTTP\n\tknative.dev/serving/pkg/queue/request_metric.go:196\nmain.proxyHandler.func1\n\tknative.dev/serving/cmd/queue/main.go:169\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.ForwardedShimHandler.func1\n\tknative.dev/serving/pkg/queue/forwarded_shim.go:100\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.(*timeoutHandler).ServeHTTP.func1\n\tknative.dev/serving/pkg/queue/timeout.go:80"}
{"level":"error","ts":"2020-06-19T18:45:56.517Z","logger":"queueproxy","caller":"network/error_handler.go:31","msg":"error reverse proxying request; sockstat: sockets: used 101884\nTCP: inuse 90 orphan 1 tw 646 alloc 2749 mem 157\nUDP: inuse 0 mem 9\nUDPLITE: inuse 0\nRAW: inuse 0\nFRAG: inuse 0 memory 0\n","commit":"804b3bb","knative.dev/key":"default/iae-timeline-predictor-default-pspwr","knative.dev/pod":"iae-timeline-predictor-default-pspwr-deployment-bc4f66df5-zhdlz","error":"context canceled","stacktrace":"knative.dev/serving/vendor/knative.dev/pkg/network.ErrorHandler.func1\n\tknative.dev/serving/vendor/knative.dev/pkg/network/error_handler.go:31\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\tnet/http/httputil/reverseproxy.go:251\nknative.dev/serving/pkg/queue.(*appRequestMetricsHandler).ServeHTTP\n\tknative.dev/serving/pkg/queue/request_metric.go:196\nmain.proxyHandler.func1\n\tknative.dev/serving/cmd/queue/main.go:169\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.ForwardedShimHandler.func1\n\tknative.dev/serving/pkg/queue/forwarded_shim.go:100\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.(*timeoutHandler).ServeHTTP.func1\n\tknative.dev/serving/pkg/queue/timeout.go:80"}
{"level":"error","ts":"2020-06-19T18:46:56.683Z","logger":"queueproxy","caller":"network/error_handler.go:31","msg":"error reverse proxying request; sockstat: sockets: used 101874\nTCP: inuse 117 orphan 1 tw 632 alloc 2778 mem 218\nUDP: inuse 0 mem 9\nUDPLITE: inuse 0\nRAW: inuse 0\nFRAG: inuse 0 memory 0\n","commit":"804b3bb","knative.dev/key":"default/iae-timeline-predictor-default-pspwr","knative.dev/pod":"iae-timeline-predictor-default-pspwr-deployment-bc4f66df5-zhdlz","error":"context canceled","stacktrace":"knative.dev/serving/vendor/knative.dev/pkg/network.ErrorHandler.func1\n\tknative.dev/serving/vendor/knative.dev/pkg/network/error_handler.go:31\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\tnet/http/httputil/reverseproxy.go:251\nknative.dev/serving/pkg/queue.(*appRequestMetricsHandler).ServeHTTP\n\tknative.dev/serving/pkg/queue/request_metric.go:196\nmain.proxyHandler.func1\n\tknative.dev/serving/cmd/queue/main.go:169\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.ForwardedShimHandler.func1\n\tknative.dev/serving/pkg/queue/forwarded_shim.go:100\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2007\nknative.dev/serving/pkg/queue.(*timeoutHandler).ServeHTTP.func1\n\tknative.dev/serving/pkg/queue/timeout.go:80"}
```

**Comment by yuzisun** [comment]
@decewei what's your custom predictor? looks like it is queue proxy is not able to connect to the main container.

**Comment by decewei** [comment]
> @decewei what's your custom predictor? looks like the queue proxy is not able to connect to the main container.

It's very similar to this one: https://github.com/kubeflow/kfserving/tree/master/docs/samples/custom/kfserving-custom-model. 

Except we are processing video with bigger model, which makes the prediction sometimes 1-2hour long. I mean this shouldn't affect the queue proxy connecting to the main container.

**Comment by yuzisun** [comment]
@decewei you can checkout the code [here](https://github.com/knative/serving/blob/c63502dea1d262a0ae838d0f8dfc75ff271dadf5/pkg/http/handler/timeout.go#L44), if the request times out it responses with 504. We are working on v1beta1 API which allows setting timeout, but 1-2 hour sounds way too long and sounds better to be a batch job.

**Comment by decewei** [comment]
> @decewei you can checkout the code [here](https://github.com/knative/serving/blob/c63502dea1d262a0ae838d0f8dfc75ff271dadf5/pkg/http/handler/timeout.go#L44), if the request times out it responses with 504. We are working on v1beta1 API which allows setting timeout, but 1-2 hour is way too long and sounds better to be a batch job.

I agree. If we can break it down to batches, we will do that. 1-2hour is an exaggeration. Probably around 30-40min depending on the input samples. There will be use case that requires longer time because it's waiting for entire video/audio segment feature segments as input.

Under those conditions, what is your suggestion in how to use KFServing for those purposes?

**Comment by yuzisun** [comment]
> > @decewei you can checkout the code [here](https://github.com/knative/serving/blob/c63502dea1d262a0ae838d0f8dfc75ff271dadf5/pkg/http/handler/timeout.go#L44), if the request times out it responses with 504. We are working on v1beta1 API which allows setting timeout, but 1-2 hour is way too long and sounds better to be a batch job.
> 
> I agree. If we can break it down to batches, we will do that. 1-2hour is an exaggeration. Probably around 30-40min depending on the input samples. There will be use case that requires longer time because it's waiting for entire video/audio segment feature segments as input.
> 
> Under those conditions, what is your suggestion in how to use KFServing for those purposes?

@decewei  KFServing is not best for this purpose, I would suggest taking a look at Kubeflow pipeline which can run video/audio segmentation as a job on the pipeline and then send the input to KFServing inference service for the prediction.

**Comment by danishsamad** [comment]
How can we change the default timeout value using  V1aplha2 specs?
I understand the "timeout" option present in the[ V1beta1PredictorSpec](https://github.com/kubeflow/kfserving/blob/master/python/kfserving/docs/V1beta1PredictorSpec.md) allows to change the default timeout value . We are currently using the KFServing 0.4.1 release and leveraging the kfserving pipeline [component ](https://github.com/kubeflow/pipelines/tree/master/components/kubeflow/kfserving ) to deploy services.

**Comment by yuzisun** [comment]
fixed in kubeflow/pipelines#5343
